File: shifai-android/app/src/main/java/com/shifai/domain/notifications/NotificationEngine.kt
package com.shifai.domain.notifications

import android.Manifest
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Build
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import androidx.core.content.ContextCompat
import java.util.*

/**
 * Smart Notification Engine — Android (S8-1 through S8-5)
 * Max 1/jour, per-category channels, anti-spam, actionable
 */
class NotificationEngine private constructor(private val context: Context) {

    companion object {
        @Volatile private var instance: NotificationEngine? = null
        fun getInstance(context: Context) = instance ?: synchronized(this) {
            instance ?: NotificationEngine(context.applicationContext).also { instance = it }
        }
    }

    enum class Category(
        val channelId: String,
        val displayName: String,
        val defaultHour: Int
    ) {
        PREDICTION("shifai_prediction", "Prédictions", 20),
        QUICK_WIN("shifai_quick_win", "Quick Wins", 9),
        EDUCATION("shifai_education", "Éducatif", 10),
        RECOMMENDATION("shifai_recommendation", "Recommandations", 8),
        REMINDER("shifai_reminder", "Rappels", 21)
    }

    private val prefs = context.getSharedPreferences("shifai_notifications", Context.MODE_PRIVATE)
    private var notifId = 1000

    // MARK: - S8-1: Create Channels (call from Application.onCreate)

    fun createChannels() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            val manager = context.getSystemService(NotificationManager::class.java)
            for (cat in Category.values()) {
                val channel = NotificationChannel(
                    cat.channelId, cat.displayName,
                    NotificationManager.IMPORTANCE_DEFAULT
                ).apply {
                    description = "Notifications ${cat.displayName} ShifAI"
                }
                manager.createNotificationChannel(channel)
            }
        }
    }

    // MARK: - S8-1: Max 1/Day Scheduler

    fun scheduleIfAllowed(category: Category, title: String, body: String, deepLink: String? = null) {
        if (!canSendToday()) return
        if (!isCategoryEnabled(category)) return
        if (isAutoStopped(category)) return
        if (isQuietHours()) return

        // Check POST_NOTIFICATIONS permission (Android 13+)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            if (ContextCompat.checkSelfPermission(context, Manifest.permission.POST_NOTIFICATIONS)
                != PackageManager.PERMISSION_GRANTED) return
        }

        val intent = Intent(Intent.ACTION_VIEW).apply {
            deepLink?.let { data = android.net.Uri.parse(it) }
            setPackage(context.packageName)
            flags = Intent.FLAG_ACTIVITY_NEW_TASK
        }
        val pendingIntent = PendingIntent.getActivity(
            context, notifId, intent,
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )

        val notification = NotificationCompat.Builder(context, category.channelId)
            .setSmallIcon(android.R.drawable.ic_dialog_info)
            .setContentTitle(title)
            .setContentText(body)
            .setContentIntent(pendingIntent)
            .setAutoCancel(true)
            .setPriority(NotificationCompat.PRIORITY_DEFAULT)
            .build()

        NotificationManagerCompat.from(context).notify(notifId++, notification)
        markSentToday()
    }

    // MARK: - S8-2: Cycle Predictions

    fun schedulePredictionNotification(daysUntilPeriod: Int, dateRange: String) {
        if (daysUntilPeriod !in 1..3) return
        scheduleIfAllowed(
            Category.PREDICTION,
            "Règles prévues dans ~$daysUntilPeriod jours",
            "Période estimée: $dateRange. Prépare-toi ☁️",
            "shifai://predictions"
        )
    }

    fun scheduleOvulationNotification(daysUntilOvulation: Int) {
        if (daysUntilOvulation !in 1..3) return
        scheduleIfAllowed(
            Category.PREDICTION,
            "Fenêtre d'ovulation dans ~$daysUntilOvulation jours",
            "Phase la plus fertile prévue bientôt 🌸",
            "shifai://predictions"
        )
    }

    // MARK: - S8-3: Quick Win & Educational

    fun scheduleQuickWinNotification(title: String, body: String) {
        val monthsUsing = prefs.getInt("months_using", 0)
        val lastTime = prefs.getLong("last_quickwin", 0L)
        val interval = if (monthsUsing <= 3) 7 * 86400_000L else 14 * 86400_000L
        if (System.currentTimeMillis() - lastTime < interval) return

        scheduleIfAllowed(Category.QUICK_WIN, title, body, "shifai://insights")
        prefs.edit().putLong("last_quickwin", System.currentTimeMillis()).apply()
    }

    fun scheduleEducationalNotification(day: Int, title: String, body: String) {
        if (day !in 4..13) return
        scheduleIfAllowed(Category.EDUCATION, title, body, "shifai://insights")
    }

    // MARK: - S8-4: Actionable Recommendations

    fun scheduleRecommendation(energyForecast: String, tip: String) {
        scheduleIfAllowed(
            Category.RECOMMENDATION,
            "☁️ $energyForecast prévue demain",
            tip, "shifai://insights"
        )
    }

    // MARK: - S8-5: Anti-Spam

    private fun canSendToday(): Boolean {
        val lastSent = prefs.getLong("last_notif_date", 0L)
        val cal = Calendar.getInstance()
        cal.timeInMillis = lastSent
        val today = Calendar.getInstance()
        return cal.get(Calendar.DAY_OF_YEAR) != today.get(Calendar.DAY_OF_YEAR) ||
                cal.get(Calendar.YEAR) != today.get(Calendar.YEAR)
    }

    private fun markSentToday() {
        prefs.edit().putLong("last_notif_date", System.currentTimeMillis()).apply()
    }

    private fun isQuietHours(): Boolean {
        val hour = Calendar.getInstance().get(Calendar.HOUR_OF_DAY)
        val start = prefs.getInt("quiet_start", 22)
        val end = prefs.getInt("quiet_end", 8)
        return if (start > end) hour >= start || hour < end else hour in start until end
    }

    private fun isAutoStopped(category: Category): Boolean {
        return prefs.getInt("ignore_${category.channelId}", 0) >= 3
    }

    fun trackIgnored(category: Category) {
        val key = "ignore_${category.channelId}"
        prefs.edit().putInt(key, prefs.getInt(key, 0) + 1).apply()
    }

    fun trackOpened(category: Category) {
        prefs.edit().putInt("ignore_${category.channelId}", 0).apply()
    }

    // MARK: - Settings

    fun isCategoryEnabled(category: Category): Boolean {
        return prefs.getBoolean("${category.channelId}_enabled", true)
    }

    fun setCategoryEnabled(category: Category, enabled: Boolean) {
        prefs.edit().putBoolean("${category.channelId}_enabled", enabled).apply()
    }
}

File: shifai-ios/ShifAI/Domain/Notifications/NotificationEngine.swift
import Foundation
import UserNotifications

// MARK: - Smart Notification Engine (S8-1 through S8-5)
// Max 1/jour, intelligentes, anti-spam, 50%+ action rate
// S8-1: Framework + scheduling
// S8-2: Cycle prediction notifications
// S8-3: Quick Win & educational
// S8-4: Actionable recommendations
// S8-5: Smart anti-spam rules

final class NotificationEngine {

    static let shared = NotificationEngine()
    private let center = UNUserNotificationCenter.current()
    private let prefs = UserDefaults.standard

    // MARK: - Categories

    enum NotificationCategory: String, CaseIterable {
        case prediction = "prediction"
        case quickWin = "quick_win"
        case education = "education"
        case recommendation = "recommendation"
        case reminder = "reminder"

        var displayName: String {
            switch self {
            case .prediction: return "Prédictions"
            case .quickWin: return "Quick Wins"
            case .education: return "Éducatif"
            case .recommendation: return "Recommandations"
            case .reminder: return "Rappels"
            }
        }

        var defaultHour: Int {
            switch self {
            case .prediction: return 20     // Evening: prepare for tomorrow
            case .quickWin: return 9        // Morning motivation
            case .education: return 10      // Mid-morning learning
            case .recommendation: return 8  // Morning planning
            case .reminder: return 21       // Evening check-in
            }
        }
    }

    // MARK: - S8-1: Permission Request (optimal timing)

    func requestPermissionIfNeeded() {
        // Don't ask during onboarding — wait for 3+ days of data
        let daysSinceInstall = prefs.integer(forKey: "days_since_install")
        guard daysSinceInstall >= 3 else { return }

        center.getNotificationSettings { settings in
            if settings.authorizationStatus == .notDetermined {
                self.center.requestAuthorization(options: [.alert, .sound, .badge]) { granted, _ in
                    if granted { self.registerCategories() }
                }
            }
        }
    }

    private func registerCategories() {
        // S8-4: Actionable recommendations
        let followAction = UNNotificationAction(
            identifier: "FOLLOW",
            title: "Oui, ajusté ✅",
            options: []
        )
        let skipAction = UNNotificationAction(
            identifier: "SKIP",
            title: "Pas cette fois",
            options: []
        )
        let feedbackCategory = UNNotificationCategory(
            identifier: "RECOMMENDATION",
            actions: [followAction, skipAction],
            intentIdentifiers: [],
            options: []
        )

        // S8-2: Prediction deep link
        let viewAction = UNNotificationAction(
            identifier: "VIEW_PREDICTION",
            title: "Voir les prédictions",
            options: [.foreground]
        )
        let predictionCategory = UNNotificationCategory(
            identifier: "PREDICTION",
            actions: [viewAction],
            intentIdentifiers: [],
            options: []
        )

        center.setNotificationCategories([feedbackCategory, predictionCategory])
    }

    // MARK: - S8-1: Max 1/Day Scheduling Engine

    func scheduleIfAllowed(category: NotificationCategory, title: String, body: String, deepLink: String? = nil) {
        // Anti-spam: check if we already sent today
        guard canSendToday() else { return }

        // Check if category is enabled
        guard isCategoryEnabled(category) else { return }

        // S8-5: Check ignore count
        guard !isAutoStopped(category) else { return }

        // S8-5: Night mode (22h-8h default)
        guard !isQuietHours() else { return }

        let content = UNMutableNotificationContent()
        content.title = title
        content.body = body
        content.sound = .default
        content.categoryIdentifier = category.rawValue.uppercased()

        if let deepLink = deepLink {
            content.userInfo = ["deep_link": deepLink]
        }

        // Schedule for preferred hour
        let hour = preferredHour(for: category)
        var dateComponents = DateComponents()
        dateComponents.hour = hour
        dateComponents.minute = 0

        let trigger = UNCalendarNotificationTrigger(dateMatching: dateComponents, repeats: false)
        let request = UNNotificationRequest(
            identifier: "\(category.rawValue)_\(UUID().uuidString)",
            content: content,
            trigger: trigger
        )

        center.add(request) { error in
            if error == nil {
                self.markSentToday()
                self.trackNotification(category: category, title: title)
            }
        }
    }

    // MARK: - S8-2: Cycle Prediction Notifications

    func schedulePredictionNotification(daysUntilPeriod: Int, dateRange: String) {
        guard daysUntilPeriod <= 3 && daysUntilPeriod > 0 else { return }

        scheduleIfAllowed(
            category: .prediction,
            title: "Règles prévues dans ~\(daysUntilPeriod) jours",
            body: "Période estimée: \(dateRange). Prépare-toi ☁️",
            deepLink: "shifai://predictions"
        )
    }

    func scheduleOvulationNotification(daysUntilOvulation: Int) {
        guard daysUntilOvulation <= 3 && daysUntilOvulation > 0 else { return }

        scheduleIfAllowed(
            category: .prediction,
            title: "Fenêtre d'ovulation dans ~\(daysUntilOvulation) jours",
            body: "Phase la plus fertile prévue bientôt 🌸",
            deepLink: "shifai://predictions"
        )
    }

    // MARK: - S8-3: Quick Win & Educational Notifications

    func scheduleQuickWinNotification(title: String, body: String) {
        // Adaptive frequency: 1x/week M1-3, 1x/2weeks after
        let monthsUsing = prefs.integer(forKey: "months_using_app")
        let lastQuickWin = prefs.object(forKey: "last_quickwin_notif") as? Date ?? .distantPast
        let interval: TimeInterval = monthsUsing <= 3 ? 7 * 86400 : 14 * 86400

        guard Date().timeIntervalSince(lastQuickWin) > interval else { return }

        scheduleIfAllowed(category: .quickWin, title: title, body: body, deepLink: "shifai://insights")
        prefs.set(Date(), forKey: "last_quickwin_notif")
    }

    func scheduleEducationalNotification(day: Int, title: String, body: String) {
        // J4-J13 daily, auto-stop J14
        guard day >= 4 && day <= 13 else { return }

        scheduleIfAllowed(category: .education, title: title, body: body, deepLink: "shifai://insights")
    }

    // MARK: - S8-4: Actionable Recommendation Notifications

    func scheduleRecommendation(energyForecast: String, tip: String) {
        scheduleIfAllowed(
            category: .recommendation,
            title: "☁️ \(energyForecast) prévue demain",
            body: tip,
            deepLink: "shifai://insights"
        )
    }

    // MARK: - S8-5: Smart Anti-Spam Rules

    private func canSendToday() -> Bool {
        let lastSent = prefs.object(forKey: "last_notification_date") as? Date
        guard let last = lastSent else { return true }
        return !Calendar.current.isDateInToday(last)
    }

    private func markSentToday() {
        prefs.set(Date(), forKey: "last_notification_date")
    }

    private func isQuietHours() -> Bool {
        let hour = Calendar.current.component(.hour, from: Date())
        let quietStart = prefs.integer(forKey: "quiet_hours_start")
        let quietEnd = prefs.integer(forKey: "quiet_hours_end")
        let start = quietStart > 0 ? quietStart : 22
        let end = quietEnd > 0 ? quietEnd : 8

        if start > end {  // Overnight (22-8)
            return hour >= start || hour < end
        } else {
            return hour >= start && hour < end
        }
    }

    private func isAutoStopped(_ category: NotificationCategory) -> Bool {
        // If user ignored 3x same type → auto-stop
        let ignoreKey = "notif_ignore_count_\(category.rawValue)"
        return prefs.integer(forKey: ignoreKey) >= 3
    }

    func trackIgnored(category: NotificationCategory) {
        let key = "notif_ignore_count_\(category.rawValue)"
        prefs.set(prefs.integer(forKey: key) + 1, forKey: key)
    }

    func trackOpened(category: NotificationCategory) {
        // Reset ignore count on open
        prefs.set(0, forKey: "notif_ignore_count_\(category.rawValue)")
    }

    // MARK: - Settings

    func isCategoryEnabled(_ category: NotificationCategory) -> Bool {
        prefs.object(forKey: "notif_\(category.rawValue)_enabled") as? Bool ?? true
    }

    func setCategoryEnabled(_ category: NotificationCategory, enabled: Bool) {
        prefs.set(enabled, forKey: "notif_\(category.rawValue)_enabled")
    }

    func preferredHour(for category: NotificationCategory) -> Int {
        let stored = prefs.integer(forKey: "notif_\(category.rawValue)_hour")
        return stored > 0 ? stored : category.defaultHour
    }

    func setPreferredHour(for category: NotificationCategory, hour: Int) {
        prefs.set(hour, forKey: "notif_\(category.rawValue)_hour")
    }

    // MARK: - Analytics

    private func trackNotification(category: NotificationCategory, title: String) {
        // Track: notification_sent event (zero PII)
        let event: [String: Any] = [
            "category": category.rawValue,
            "timestamp": ISO8601DateFormatter().string(from: Date())
        ]
        // TODO: Send to Plausible analytics
        _ = event
    }

    func handleAction(identifier: String, category: String) {
        switch identifier {
        case "FOLLOW":
            // recommendation_followed event
            if let cat = NotificationCategory(rawValue: category.lowercased()) {
                trackOpened(category: cat)
            }
        case "SKIP":
            if let cat = NotificationCategory(rawValue: category.lowercased()) {
                trackIgnored(category: cat)
            }
        default:
            break
        }
    }
}
